core/src/main/scala/kafka/server/AlterIsrManager.scala

```
@@ -250,23 +260,35 @@ class DefaultAlterIsrManager(
         val partitionResponses: mutable.Map[TopicPartition, Either[Errors, LeaderAndIsr]] =
           new mutable.HashMap[TopicPartition, Either[Errors, LeaderAndIsr]]()
         data.topics.forEach { topic =>
-          topic.partitions().forEach(partition => {
+          topic.partitions().forEach { partition =>
             val tp = new TopicPartition(topic.name, partition.partitionIndex)
-            val error = Errors.forCode(partition.errorCode())
+            val apiError = Errors.forCode(partition.errorCode())
             debug(s"Controller successfully handled AlterIsr request for $tp: $partition")
-            if (error == Errors.NONE) {
-              val newLeaderAndIsr = new LeaderAndIsr(partition.leaderId, partition.leaderEpoch,
-                partition.isr.asScala.toList.map(_.toInt), partition.currentIsrVersion)
-              partitionResponses(tp) = Right(newLeaderAndIsr)
+            if (apiError == Errors.NONE) {
+              try {
+                partitionResponses(tp) = Right(
+                  LeaderAndIsr(
+                    partition.leaderId,
+                    partition.leaderEpoch,
+                    partition.isr.asScala.toList.map(_.toInt),
+                    LeaderRecoveryState.of(partition.leaderRecoveryState),
+                    partition.partitionEpoch
+                  )
+                )
+              } catch {
+                case e: IllegalArgumentException =>
```

It might be cleaner to let `LeaderRecoveryState.of` return `Optional<LeaderRecoveryState>`

---

Yeah. I went back and forth of this. There are some code paths were we want to throw an exception as in most cases this means a bug in Kafka. I think this was the only case where we want to handle this exception. I'll take a look again.

---

Okay. We have both: `of` which throws and `optionalOf` which returns an `Optional`.

---

Yeah, I think the main thing I didn't like is that `IllegalArgumentException` is such a generic error. So using it to infer a particular case seemed a little slippery.